File: MP4/src/main/java/edu/illinois/storm/RandomSentenceSpout.java
package edu.illinois.storm;

import java.util.Map;
import java.util.Random;
import org.apache.storm.spout.SpoutOutputCollector;
import org.apache.storm.task.TopologyContext;
import org.apache.storm.topology.OutputFieldsDeclarer;
import org.apache.storm.topology.base.BaseRichSpout;
import org.apache.storm.tuple.Fields;
import org.apache.storm.tuple.Values;
import org.apache.storm.utils.Utils;

/** a spout that randomly generate sentence from a set of sentences */
public class RandomSentenceSpout extends BaseRichSpout {
  SpoutOutputCollector _collector;
  Random _rand;
  String[] sentences;

  public void open(Map conf, TopologyContext context, SpoutOutputCollector collector) {
    _collector = collector;
    _rand = new Random();
    sentences =
        new String[] {
          "the cow jumped over the moon",
          "an apple a day keeps the doctor away",
          "four score and seven years ago",
          "snow white and the seven dwarfs",
          "i am at two with nature"
        };
  }

  public void nextTuple() {
    Utils.sleep(200);
    /* ----------------------TODO-----------------------
    Task: randomly generate sentence from sentences string array
    ------------------------------------------------- */
    String sentence = sentences[_rand.nextInt(sentences.length)];
    _collector.emit(new Values(sentence));
		// End
  }

  public void declareOutputFields(OutputFieldsDeclarer declarer) {
    /* ----------------------TODO-----------------------
    Task: declare output fields
    ------------------------------------------------- */
    declarer.declare(new Fields("word"));
		// End
  }
}

File: MP4_python/multilang/resources/random_sentence_spout.py
import storm
import random
import time

# Define some sentences
SENTENCES = """
the cow jumped over the moon
an apple a day keeps the doctor away
four score and seven years ago
snow white and the seven dwarfs
i am at two with nature
""".strip().split('\n')


class SentenceSpout(storm.Spout):
    def initialize(self, conf, context):
        self._conf = conf
        self._context = context

        storm.logInfo("Spout instance starting...")

    # Process the next tuple
    def nextTuple(self):
        time.sleep(0.2)
        # TODO
        # Task: randomly generate sentence from sentences string array
        # Note: only generate one sentence in this function
        sentence = random.choice(SENTENCES)
        # storm.logInfo("Emiting %s" % sentence)
        storm.emit([sentence])
        # pass - just a placeholder
        # End


# Start the spout when it's invoked
SentenceSpout().run()
